File: samples/csharp/assistants/hospitality-assistant/VirtualAssistantSample/Dialogs/OnboardingDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Globalization;
using System.Threading;
using System.Threading.Tasks;
using Luis;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Solutions.Responses;
using Microsoft.Extensions.DependencyInjection;
using VirtualAssistantSample.Models;
using VirtualAssistantSample.Services;

namespace VirtualAssistantSample.Dialogs
{
    // Example onboarding dialog to initial user profile information.
    public class OnboardingDialog : ComponentDialog
    {
        private BotServices _services;
        private LocaleTemplateEngineManager _templateEngine;
        private IStatePropertyAccessor<UserProfileState> _accessor;

        public OnboardingDialog(
            IServiceProvider serviceProvider,
            IBotTelemetryClient telemetryClient)
            : base(nameof(OnboardingDialog))
        {
            _templateEngine = serviceProvider.GetService<LocaleTemplateEngineManager>();

            var userState = serviceProvider.GetService<UserState>();
            _accessor = userState.CreateProperty<UserProfileState>(nameof(UserProfileState));
            _services = serviceProvider.GetService<BotServices>();

            var onboarding = new WaterfallStep[]
            {
                AskForName,
                FinishOnboardingDialog,
            };

            // To capture built-in waterfall dialog telemetry, set the telemetry client
            // to the new waterfall dialog and add it to the component dialog
            TelemetryClient = telemetryClient;
            AddDialog(new WaterfallDialog(nameof(onboarding), onboarding) { TelemetryClient = telemetryClient });
            AddDialog(new TextPrompt(DialogIds.NamePrompt));
        }

        public async Task<DialogTurnResult> AskForName(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            var state = await _accessor.GetAsync(sc.Context, () => new UserProfileState());

            if (!string.IsNullOrEmpty(state.Name))
            {
                return await sc.NextAsync(state.Name);
            }
            else
            {
                return await sc.PromptAsync(DialogIds.NamePrompt, new PromptOptions()
                {
                    Prompt = _templateEngine.GenerateActivityForLocale("NamePrompt"),
                });
            }
        }

        public async Task<DialogTurnResult> FinishOnboardingDialog(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            var userProfile = await _accessor.GetAsync(sc.Context, () => new UserProfileState());
            var name = (string)sc.Result;

            var generalResult = sc.Context.TurnState.Get<GeneralLuis>(StateProperties.GeneralResult);
            if (generalResult == null)
            {
                var localizedServices = _services.GetCognitiveModels();
                generalResult = await localizedServices.LuisServices["General"].RecognizeAsync<GeneralLuis>(sc.Context, cancellationToken);
            }

            (var generalIntent, var generalScore) = generalResult.TopIntent();
            if (generalIntent == GeneralLuis.Intent.ExtractName && generalScore > 0.5)
            {
                if (generalResult.Entities.PersonName_Any != null)
                {
                    name = generalResult.Entities.PersonName_Any[0];
                }
                else if (generalResult.Entities.personName != null)
                {
                    name = generalResult.Entities.personName[0];
                }
            }

            // Captialize name
            userProfile.Name = CultureInfo.CurrentCulture.TextInfo.ToTitleCase(name.ToLower());

            await _accessor.SetAsync(sc.Context, userProfile, cancellationToken);

            await sc.Context.SendActivityAsync(_templateEngine.GenerateActivityForLocale("HaveNameMessage", userProfile));

            return await sc.EndDialogAsync();
        }

        private class DialogIds
        {
            public const string NamePrompt = "namePrompt";
        }

        private class StateProperties
        {
            public const string GeneralResult = "generalResult";
        }
    }
}

File: templates/typescript/samples/sample-assistant/src/dialogs/onboardingDialog.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */
import { StatePropertyAccessor, RecognizerResult } from 'botbuilder';
import {
    ComponentDialog,
    DialogTurnResult,
    TextPrompt,
    WaterfallDialog,
    WaterfallStepContext, 
    WaterfallStep } from 'botbuilder-dialogs';
import { IUserProfileState } from '../models/userProfileState';
import { StateProperties } from '../models/stateProperties';
import { BotServices } from '../services/botServices';
import { LocaleTemplateManager } from 'bot-solutions';
import { LuisRecognizer } from 'botbuilder-ai';

enum DialogIds {
    NamePrompt = 'namePrompt',
}

// Example onboarding dialog to initial user profile information.
export class OnboardingDialog extends ComponentDialog {
    private readonly services: BotServices;
    private readonly templateManager: LocaleTemplateManager;
    private readonly accessor: StatePropertyAccessor<IUserProfileState>;

    public constructor(
        accessor: StatePropertyAccessor<IUserProfileState>,
        services: BotServices,
        templateManager: LocaleTemplateManager) {
        super(OnboardingDialog.name);
        this.templateManager = templateManager;

        this.accessor = accessor;
        this.services = services;

        const onboarding: WaterfallStep[] = [
            this.askForName.bind(this),
            this.finishOnboardingDialog.bind(this)
        ];

        this.addDialog(new WaterfallDialog(OnboardingDialog.name, onboarding));
        this.addDialog(new TextPrompt(DialogIds.NamePrompt));
    }

    public async askForName(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        const state: IUserProfileState = await this.accessor.get(sc.context, { name: ''});

        if (state.name !== undefined && state.name.trim().length > 0) {
            return await sc.next(state.name);
        }
        
        return await sc.prompt(DialogIds.NamePrompt, {
            prompt: this.templateManager.generateActivityForLocale('NamePrompt', sc.context.activity.locale),
        });
    }

    public async finishOnboardingDialog(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        const userProfile: IUserProfileState = await this.accessor.get(sc.context, { name: '' });
        let name: string = sc.result as string;

        let generalResult: RecognizerResult | undefined = sc.context.turnState.get(StateProperties.GeneralResult);
        if (generalResult === undefined) {
            const localizedServices = this.services.getCognitiveModels(sc.context.activity.locale as string);
            generalResult = await localizedServices.luisServices.get('General')?.recognize(sc.context);
            if (generalResult) {
                sc.context.turnState.set(StateProperties.GeneralResult, generalResult);
            }
        }

        if (generalResult !== undefined) {
            const intent: string = LuisRecognizer.topIntent(generalResult);
            if (intent === 'ExtractName' && generalResult.intents[intent].score > 0.5) {
                if (generalResult.entities['PersonName_Any'] !== undefined) {
                    name = generalResult.entities['PersonName_Any'];
                } else if (generalResult.entities['personName'] !== undefined) {
                    name = generalResult.entities['personName'];
                }
            }
        }

        // Capitalize name
        userProfile.name = name.toLowerCase()
            .split(' ')
            .map((word: string): string => word.charAt(0)
                .toUpperCase() + word.substring(1))
            .join(' ');

        await this.accessor.set(sc.context, userProfile);

        await sc.context.sendActivity(this.templateManager.generateActivityForLocale('HaveNameMessage', sc.context.activity.locale, userProfile));

        return await sc.endDialog();
    }
}